ttnn/cpp/ttnn/operations/data_movement.hpp

```diff
@@ -315,9 +319,117 @@ struct RepeatInterleave {
     }
 };
 
+struct Pad {
+    static inline const std::array<TensorSchema, 1> input_tensor_schemas() {
+        return {ttnn::TensorSchema{
+            2,  // min rank
+            4,  // max rank
+            {ttnn::bfloat16, ttnn::bfloat8_b, ttnn::uint16, ttnn::int32, ttnn::uint32},
+            {ttnn::TILE_LAYOUT},
+            true,   // can_be_on_device
+            false,  // can_be_on_cpu
+            false,  // can_be_scalar
+            false   // is_optional}
+        }};
+    }
+
+    template <typename... Args>
+    static auto input_tensors_to_validate(const ttnn::Tensor& input_tensor, Args&&... args) {
+        return std::make_tuple(input_tensor);
+    }
+
+    static ttnn::Tensor execute_on_worker_thread(
+        const ttnn::Tensor& input_tensor,
+        const std::vector<std::pair<size_t, size_t>>& padding,
+        const float value,
+        const std::optional<MemoryConfig>& memory_config_arg) {
+
+        const int original_rank = input_tensor.get_shape().rank();
+        TT_FATAL(
+            padding.size() == original_rank,
+            "ttnn.pad: padding must be the same length as the input tensor rank");
+        TT_FATAL(
+            input_tensor.get_layout() != ttnn::ROW_MAJOR_LAYOUT,
+            "ttnn.pad: row-major tensors have to use fallback because the kernel currently causes a PCC error");
+
+        const bool padding_has_negative_value = std::any_of(padding.begin(), padding.end(), [](const auto& p) { return p.first < 0 || p.second < 0; });
+        TT_FATAL(
+            !padding_has_negative_value,
+            "ttnn.pad: padding must be non-negative");
+
+        // Unsqueeze Tensor to 4D if it is not already
+        ttnn::Tensor input_tensor_4D = input_tensor;
+        auto padding_4D = padding;
+        if (original_rank < 4) {
+            input_tensor_4D = ttnn::unsqueeze_to_4D(input_tensor);
+            padding_4D.insert(padding_4D.begin(), 4 - original_rank, {0, 0});
+        }
+        auto input_shape_with_tile_padding = input_tensor_4D.get_shape().with_tile_padding();
+        std::vector<uint32_t> output_padded_shape(padding_4D.size());
+        for(int i = 0; i < padding_4D.size(); i++) {
+            output_padded_shape[i] = input_shape_with_tile_padding[i] + padding_4D[i].second;
+        }
+
+        // Due to the strangeness of tt::tt_metal::pad, we need to split front and back pad
+        // Front will be passed separately. And pad_back is retrieved -> output_padded_shape - pad_front
+        auto memory_config = memory_config_arg.value_or(input_tensor.memory_config());
+        auto pad_front = padding_4D | std::views::transform([](const auto& p) { return p.first; });
+        auto pad_back = padding_4D | std::views::transform([](const auto& p) { return p.second; });
+
+        const bool padding_starts_from_zero = std::accumulate(pad_front.begin(), pad_front.end(), 0) == 0;
```

Comment:
not sure  I understood this case.  It would be nice if you can point me out to the test which covers/

Reply:
We do not support non-zero front padding right now